src/methods/destructuring.js

```diff
@@ -0,0 +1,434 @@
+        throw e;
+      }
+
+      // d. If next is false, set iteratorRecord.[[Done]] to true.
+      if (next === false) {
+        iteratorRecord.$Done = true;
+        // Normally this assignment would be done in step 3, but we do it
+        // here so that Flow knows `value` will always be initialized by step 4.
+        value = new UndefinedValue(realm);
+      } else { // e. Else,
+        // i. Let value be IteratorValue(next).
+        try {
+          value = IteratorValue(realm, next);
+        } catch (e) {
+          // ii. If value is an abrupt completion, set iteratorRecord.[[Done]] to true.
+          if (e instanceof AbruptCompletion) {
+            iteratorRecord.$Done = true;
+          }
+          // iii. ReturnIfAbrupt(v).
+          throw e;
+        }
+      }
+    } else { // 3. If iteratorRecord.[[Done]] is true, let value be undefined.
+      value = new UndefinedValue(realm);
+    }
+
+    let v;
+
+    // 4. If Initializer is present and value is undefined, then
+    if (Initializer && value instanceof UndefinedValue) {
+      // a. Let defaultValue be the result of evaluating Initializer.
+      let defaultValue = env.evaluate(Initializer, strictCode);
+
+      // b. Let v be ? GetValue(defaultValue).
+      v = GetValue(realm, defaultValue);
+    } else { // 5. Else, let v be value.
+      v = value;
+    }
+
+    // 6. If DestructuringAssignmentTarget is an ObjectLiteral or an ArrayLiteral, then
+    //
+    // The spec assumes we haven't yet distinguished between literals and
+    // patterns, but our parser does that work for us. That means we check for
+    // "*Pattern" instead of "*Literal" like the spec text suggests.
+    if (DestructuringAssignmentTarget.type === "ObjectPattern" ||
+        DestructuringAssignmentTarget.type === "ArrayPattern") {
+      // a. Let nestedAssignmentPattern be the parse of the source text corresponding to DestructuringAssignmentTarget using either AssignmentPattern or AssignmentPattern[Yield] as the goal symbol depending upon whether this AssignmentElement has the [Yield] parameter.
+      let nestedAssignmentPattern = DestructuringAssignmentTarget;
+
+      // b. Return the result of performing DestructuringAssignmentEvaluation of nestedAssignmentPattern with v as the argument.
+      DestructuringAssignmentEvaluation(realm, nestedAssignmentPattern, v, strictCode, env);
+      continue;
+    }
+
+    // We know `lref` exists because of how the algorithm is setup, but tell
+    // Flow that `lref` exists with an `invariant()`.
+    invariant(lref);
+
+    // 7. If Initializer is present and value is undefined and IsAnonymousFunctionDefinition(Initializer) and IsIdentifierRef of DestructuringAssignmentTarget are both true, then
+    if (Initializer &&
+        value instanceof UndefinedValue &&
+        IsAnonymousFunctionDefinition(realm, Initializer) &&
+        IsIdentifierRef(realm, DestructuringAssignmentTarget) &&
+        v instanceof ObjectValue) {
```

Comment:
`v` should always be an `ObjectValue` if the rest of the `if` is true, but Flow doesn’t know that so that’s why the `v instanceof ObjectValue` is there. It could be an invariant if you would rather.